File: src/livekit_agent_simulator/audio/degradation.py
"""Audio degradation effects for the simulated caller's mic.

Applies packet-loss / echo / phone-band / static degradation to the caller PCM
so the agent under test is stressed with the imperfect audio real callers hit —
the "call your users actually make" gap LangWatch Scenario exposes in its
``effects`` pipeline (noise / prosody / quality). lks keeps the surface small:
knobs under ``Persona.speech_conditions.effects``, each a pure per-frame PCM
transform, deterministic, no new runtime deps.

Effects are applied AFTER mixing (speech + noise → one frame) in
``ParallelMicMixer._pop_frame``, so the recorder captures exactly what the
agent would hear. They run on PCM16 mono at the caller output rate (24 kHz).
"""

from __future__ import annotations

import array
import random
from typing import Any, Callable

# A post-mix frame transform: PCM16 mono bytes in, PCM16 mono bytes out.
EffectFn = Callable[[bytes], bytes]


def _seeded_rng(*parts: Any) -> random.Random:
    """Build a deterministic RNG from the effect's config.

    The same effect config (e.g. ``packet_loss probability=0.05 chunk_ms=20``)
    always yields the same dropout pattern across runs, so a scenario with
    degradation is reproducible in CI. Different configs seed differently.
    """
    seed = "|".join(str(p) for p in parts)
    return random.Random(seed)


def _pcm_to_samples(pcm: bytes) -> array.array:
    if not pcm:
        return array.array("h")
    if len(pcm) % 2:
        pcm = pcm[:-1]
    a = array.array("h")
    a.frombytes(pcm)
    return a


def _samples_to_pcm(samples: array.array) -> bytes:
    return samples.tobytes()


def _dropout(samples: array.array, chunk_samples: int, probability: float, rng: random.Random) -> array.array:
    """Zero out random ``chunk_samples`` windows at ``probability``.

    ``chunk_samples`` is clamped to the frame length so short mixer frames
    (e.g. 10ms @ 24 kHz = 240 samples vs a 20ms/480-sample window) still
    experience dropouts instead of silently passing through.
    """
    if probability <= 0 or not samples:
        return samples
    chunk = max(1, min(chunk_samples, len(samples)))
    out = array.array("h", samples)
    i = 0
    while i < len(out):
        if rng.random() < probability:
            out[i : i + chunk] = array.array("h", [0] * min(chunk, len(out) - i))
        i += chunk
    return out


def packet_loss(probability: float = 0.05, chunk_ms: int = 20, sample_rate: int = 24_000) -> EffectFn:
    """Simulate network dropouts: zero random ~20ms windows (seeded, reproducible)."""
    chunk = max(1, (sample_rate * chunk_ms) // 1000)
    rng = _seeded_rng("packet_loss", probability, chunk)

    def _apply(pcm: bytes) -> bytes:
        if probability <= 0 or not pcm:
            return pcm
        return _samples_to_pcm(_dropout(_pcm_to_samples(pcm), chunk, probability, rng))

    return _apply


def breaking_up(probability: float = 0.2, sample_rate: int = 24_000) -> EffectFn:
    """Intermittent connection: heavier dropouts than packet_loss (LangWatch parity, seeded)."""
    chunk = (sample_rate * 100) // 1000  # 100ms windows
    rng = _seeded_rng("breaking_up", probability, chunk)

    def _apply(pcm: bytes) -> bytes:
        if not pcm:
            return pcm
        return _samples_to_pcm(_dropout(_pcm_to_samples(pcm), chunk, probability, rng))

    return _apply


def echo(delay_ms: int = 200, decay: float = 0.5, sample_rate: int = 24_000) -> EffectFn:
    """Overlay a delayed, attenuated copy of the signal (acoustic echo)."""
    delay_samples = (sample_rate * delay_ms) // 1000
    if delay_samples <= 0:
        raise ValueError(f"echo delay_ms must be positive (got {delay_ms})")

    def _apply(pcm: bytes) -> bytes:
        if not pcm:
            return pcm
        samples = _pcm_to_samples(pcm)
        out = array.array("h", samples)
        for i, s in enumerate(samples):
            src = i - delay_samples
            if src >= 0:
                v = int(out[i]) + int(round(samples[src] * decay))
                if v > 32767:
                    v = 32767
                elif v < -32768:
                    v = -32768
                out[i] = v
        return _samples_to_pcm(out)

    return _apply


def phone_quality(sample_rate: int = 24_000) -> EffectFn:
    """Mimic a phone line: 4 kHz bandpass + mild amplitude compression.

    Simple two-pass lowpass (moving average) approximates the 300–3.4 kHz
    bandpass without heavy DSP deps. Compression gently attenuates loud peaks.
    """
    # ~4 kHz cutoff with a short moving-average window at 24 kHz → phone-ish.
    window = max(1, sample_rate // 6000)
    if sample_rate // 6000 < 1:
        window = 1

    def _apply(pcm: bytes) -> bytes:
        if not pcm:
            return pcm
        samples = _pcm_to_samples(pcm)
        out = array.array("h", [0] * len(samples))
        acc = 0
        for i, s in enumerate(samples):
            acc += int(s)
            if i >= window:
                acc -= int(samples[i - window])
            avg = acc // window
            # mild compression toward the mid-band
            v = avg
            if v > 8000:
                v = 8000 + (v - 8000) // 2
            elif v < -8000:
                v = -8000 - (v + 8000) // 2
            out[i] = v
        return _samples_to_pcm(out)

    return _apply


def static(intensity: float = 0.05) -> EffectFn:
    """Overlay white-noise static at the given intensity (fraction of full scale, seeded)."""
    if not 0.0 <= intensity <= 1.0:
        raise ValueError(f"static intensity must be in [0.0, 1.0] (got {intensity})")
    rng = _seeded_rng("static", intensity)

    def _apply(pcm: bytes) -> bytes:
        if not pcm:
            return pcm
        samples = _pcm_to_samples(pcm)
        out = array.array("h", samples)
        for i in range(len(out)):
            noise = int(round(rng.gauss(0, 32767 * intensity)))
            v = out[i] + noise
            if v > 32767:
                v = 32767
            elif v < -32768:
                v = -32768
            out[i] = v
        return _samples_to_pcm(out)

    return _apply


# name → factory
_EFFECTS: dict[str, Callable[[dict[str, Any]], EffectFn]] = {
    "packet_loss": lambda kw: packet_loss(
        probability=float(kw.get("probability", 0.05)),
        chunk_ms=int(kw.get("chunk_ms", 20)),
    ),
    "breaking_up": lambda kw: breaking_up(probability=float(kw.get("probability", 0.2))),
    "echo": lambda kw: echo(
        delay_ms=int(kw.get("delay_ms", 200)),
        decay=float(kw.get("decay", 0.5)),
    ),
    "phone_quality": lambda kw: phone_quality(),
    "static": lambda kw: static(intensity=float(kw.get("intensity", 0.05))),
}

SUPPORTED_EFFECTS = tuple(sorted(_EFFECTS))


def effects_spec_from_persona(persona: dict[str, Any] | None) -> dict[str, Any] | None:
    """Extract ``speech_conditions.effects`` from a persona dict ('' aliases included)."""
    if not isinstance(persona, dict):
        return None
    sc = persona.get("speech_conditions") or persona.get("speechConditions") or {}
    if not isinstance(sc, dict):
        return None
    return sc.get("effects")


def resolve_audio_effects(persona: dict[str, Any] | None) -> list[EffectFn]:
    """Build the ordered effect chain from ``Persona.speech_conditions.effects``.

    Accepts the persona dict (matches ``resolve_voice_gain``) and reads
    ``speech_conditions.effects``, which is either a dict mapping
    ``{name: kwargs-or-bool}`` or a list of names. Unsupported names raise so
    a typo fails fast at validate time, not mid-run.
    """
    spec = effects_spec_from_persona(persona)
    if not spec:
        return []
    if isinstance(spec, (list, tuple)):
        items = {str(n): {} for n in spec}
    elif isinstance(spec, dict):
        items = spec
    else:
        raise ValueError(
            "Persona.speech_conditions.effects must be a list or dict "
            f"(got {type(spec).__name__})"
        )

    chain: list[EffectFn] = []
    for name, cfg in items.items():
        key = str(name).strip().lower()
        if key not in _EFFECTS:
            raise ValueError(
                f"Unknown audio effect {name!r}; supported: {', '.join(SUPPORTED_EFFECTS)}"
            )
        kwargs: dict[str, Any] = {}
        if isinstance(cfg, dict):
            kwargs = cfg
        elif cfg not in (None, True):
            raise ValueError(f"audio effect {name!r} config must be a dict, true, or absent")
        chain.append(_EFFECTS[key](kwargs))
    return chain

File: src/livekit_agent_simulator_rust/crates/lks-livekit/src/degradation.rs
//! Audio degradation effects (port of `audio/degradation.py`) — packet-loss,
//! breaking-up, echo, phone-quality, static. Deterministic (seeded), applied
//! per PCM frame so the agent hears imperfect audio like a real caller.

use rand::{rngs::StdRng, RngExt, SeedableRng};

pub const SUPPORTED_EFFECTS: [&str; 5] = [
    "breaking_up",
    "echo",
    "packet_loss",
    "phone_quality",
    "static",
];

type Samples = Vec<i16>;

/// One PCM-frame effect in the chain (packet-loss / echo / …).
pub type PcmEffect = Box<dyn Fn(&[u8]) -> Vec<u8> + Send>;
pub type PcmEffectChain = Vec<PcmEffect>;

fn pcm_to_samples(pcm: &[u8]) -> Samples {
    let mut s: Vec<i16> = Vec::with_capacity(pcm.len() / 2);
    let mut i = 0;
    while i + 1 < pcm.len() {
        s.push(i16::from_le_bytes([pcm[i], pcm[i + 1]]));
        i += 2;
    }
    s
}

fn samples_to_pcm(s: &Samples) -> Vec<u8> {
    let mut out = Vec::with_capacity(s.len() * 2);
    for v in s {
        out.extend_from_slice(&v.to_le_bytes());
    }
    out
}

fn dropout(samples: &mut Samples, chunk_samples: usize, probability: f64, rng: &mut StdRng) {
    if probability <= 0.0 || samples.is_empty() {
        return;
    }
    let chunk = chunk_samples.clamp(1, samples.len());
    let mut i = 0;
    while i < samples.len() {
        if rng.random::<f64>() < probability {
            let n = chunk.min(samples.len() - i);
            for s in samples.iter_mut().skip(i).take(n) {
                *s = 0;
            }
        }
        i += chunk;
    }
}

fn seed_rng(parts: &[&str]) -> StdRng {
    let seed = parts.join("|");
    // Hash the string into a [u8; 32] seed (deterministic across runs).
    let mut h = 0xcbf29ce484222325u64;
    for b in seed.bytes() {
        h ^= b as u64;
        h = h.wrapping_mul(0x100000001b3);
    }
    let mut bytes = [0u8; 32];
    for (i, b) in bytes.iter_mut().enumerate() {
        h = h.wrapping_mul(0x100000001b3) ^ (i as u64).wrapping_add(1);
        *b = (h >> ((i % 8) * 8)) as u8;
    }
    StdRng::from_seed(bytes)
}

/// Build the ordered effect chain from `speech_conditions.effects`.
pub fn resolve_audio_effects(spec: Option<&serde_json::Value>) -> Result<PcmEffectChain, String> {
    let mut chain: PcmEffectChain = Vec::new();
    let Some(spec) = spec else { return Ok(chain) };
    let entries: Vec<(String, serde_json::Value)> = match spec {
        serde_json::Value::Object(m) => m.iter().map(|(k, v)| (k.clone(), v.clone())).collect(),
        serde_json::Value::Array(a) => a
            .iter()
            .filter_map(|v| {
                v.as_str()
                    .map(|s| (s.to_string(), serde_json::Value::Bool(true)))
            })
            .collect(),
        _ => return Ok(chain),
    };
    for (name, kw) in entries {
        let kwargs = kw.as_object().cloned().unwrap_or_default();
        match name.as_str() {
            "packet_loss" => {
                let probability = kwargs
                    .get("probability")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.05);
                let chunk_ms = kwargs
                    .get("chunk_ms")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(20) as usize;
                let chunk = (24_000usize * chunk_ms).max(1) / 1000;
                let rng = std::cell::RefCell::new(seed_rng(&[
                    "packet_loss",
                    &probability.to_string(),
                    &chunk.to_string(),
                ]));
                chain.push(Box::new(move |pcm: &[u8]| {
                    if probability <= 0.0 || pcm.is_empty() {
                        return pcm.to_vec();
                    }
                    let mut s = pcm_to_samples(pcm);
                    dropout(&mut s, chunk, probability, &mut rng.borrow_mut());
                    samples_to_pcm(&s)
                }));
            }
            "breaking_up" => {
                let probability = kwargs
                    .get("probability")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.2);
                let chunk = 24_000 * 100 / 1000;
                let rng = std::cell::RefCell::new(seed_rng(&[
                    "breaking_up",
                    &probability.to_string(),
                    &chunk.to_string(),
                ]));
                chain.push(Box::new(move |pcm: &[u8]| {
                    if pcm.is_empty() {
                        return pcm.to_vec();
                    }
                    let mut s = pcm_to_samples(pcm);
                    dropout(&mut s, chunk, probability, &mut rng.borrow_mut());
                    samples_to_pcm(&s)
                }));
            }
            "echo" => {
                let delay_ms_raw = kwargs
                    .get("delay_ms")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(200);
                if delay_ms_raw <= 0 {
                    return Err(format!(
                        "echo delay_ms must be positive (got {delay_ms_raw})"
                    ));
                }
                let delay_ms = delay_ms_raw as usize;
                let decay = kwargs.get("decay").and_then(|v| v.as_f64()).unwrap_or(0.5);
                let delay_samples = 24_000usize * delay_ms / 1000;
                chain.push(Box::new(move |pcm: &[u8]| {
                    if pcm.is_empty() {
                        return pcm.to_vec();
                    }
                    let samples = pcm_to_samples(pcm);
                    let mut out = samples.clone();
                    for i in 0..samples.len() {
                        if i >= delay_samples {
                            let v = out[i] as f64 + (samples[i - delay_samples] as f64 * decay);
                            out[i] = v.round().clamp(-32768.0, 32767.0) as i16;
                        }
                    }
                    samples_to_pcm(&out)
                }));
            }
            "phone_quality" => {
                let window = 24_000usize / 6000;
                chain.push(Box::new(move |pcm: &[u8]| {
                    if pcm.is_empty() {
                        return pcm.to_vec();
                    }
                    let samples = pcm_to_samples(pcm);
                    let mut out = vec![0i16; samples.len()];
                    let mut acc: i64 = 0;
                    for (i, &s) in samples.iter().enumerate() {
                        acc += s as i64;
                        if i >= window {
                            acc -= samples[i - window] as i64;
                        }
                        let mut avg = (acc / window as i64) as i32;
                        if avg > 8000 {
                            avg = 8000 + (avg - 8000) / 2;
                        } else if avg < -8000 {
                            avg = -8000 - (avg + 8000) / 2;
                        }
                        out[i] = avg.clamp(-32768, 32767) as i16;
                    }
                    samples_to_pcm(&out)
                }));
            }
            "static" => {
                let intensity = kwargs
                    .get("intensity")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.05);
                if !(0.0..=1.0).contains(&intensity) {
                    return Err(format!(
                        "static intensity must be in [0.0, 1.0] (got {intensity})"
                    ));
                }
                let rng = std::cell::RefCell::new(seed_rng(&["static", &intensity.to_string()]));
                chain.push(Box::new(move |pcm: &[u8]| {
                    if pcm.is_empty() {
                        return pcm.to_vec();
                    }
                    let mut out = pcm_to_samples(pcm);
                    let amp = (32767.0 * intensity).round();
                    for v in out.iter_mut() {
                        // Gaussian-ish via two uniforms (Box-Muller-lite).
                        let u: f64 = rng.borrow_mut().random::<f64>();
                        let noise = (u - 0.5) * 2.0 * amp;
                        let nv = *v as f64 + noise;
                        *v = nv.round().clamp(-32768.0, 32767.0) as i16;
                    }
                    samples_to_pcm(&out)
                }));
            }
            other => {
                return Err(format!(
                    "unknown audio effect: {other} (supported: {})",
                    SUPPORTED_EFFECTS.join(", ")
                ));
            }
        }
    }
    Ok(chain)
}

/// Apply the effect chain to one PCM frame.
pub fn apply_effects(chain: &PcmEffectChain, pcm: &[u8]) -> Vec<u8> {
    let mut out = pcm.to_vec();
    for effect in chain {
        out = effect(&out);
    }
    out
}
